### PR TITLE
feat: add support for user defined generic mapping methods

### PR DESCRIPTION
**Pull request by latonz**
Closes https://github.com/riok/mapperly/issues/357

### CODE REVIEW

File: src/Riok.Mapperly/Descriptors/UserMethodMappingExtractor.cs [hunk 1]
```diff
@@ -107,6 +109,25 @@ bool isStatic
...
+                ctx.MapperConfiguration.UseReferenceHandling,
+                ctx.Types.PreserveReferenceHandler,
+                GetTypeSwitchNullArm(methodSymbol, parameters, typeParameters),
+                ctx.Compilation.ObjectType
+            );
+        }
+
+        if (methodSymbol.IsGenericMethod)
+        {
+            ctx.ReportDiagnostic(DiagnosticDescriptors.UnsupportedMappingMethodSignature, methodSymbol, methodSymbol.Name);
```
Comment (CommonGuy) [inline src/Riok.Mapperly/Descriptors/UserMethodMappingExtractor.cs:127]: Is there a test for this? Codecov seems to complain

File: test/Riok.Mapperly.IntegrationTests/_snapshots/NET_48/StaticMapperTest.SnapshotGeneratedSource.verified.cs [hunk 2]
```diff
@@ -374,6 +375,26 @@ public static partial class StaticTestMapper
             };
         }
 
+        public static partial TTarget MapGeneric<TSource, TTarget>(TSource source)
+        {
+            return source switch
+            {
+                global::Riok.Mapperly.IntegrationTests.Models.TestEnum x when typeof(TTarget).IsAssignableFrom(typeof(global::Riok.Mapperly.IntegrationTests.Dto.TestEnumDtoByName)) => (TTarget)(object)MapToEnumDtoByName(x),
```
Comment (CommonGuy) [inline test/Riok.Mapperly.IntegrationTests/_snapshots/NET_48/StaticMapperTest.SnapshotGeneratedSource.verified.cs:382]: Is the extra cast to `object` needed before casting it to `TTarget`?

Comment (latonz) [reply test/Riok.Mapperly.IntegrationTests/_snapshots/NET_48/StaticMapperTest.SnapshotGeneratedSource.verified.cs:382]: Yep it is 🤷

File: src/Riok.Mapperly/Descriptors/UserMethodMappingExtractor.cs (hunk 1 above)
Comment (latonz) [reply src/Riok.Mapperly/Descriptors/UserMethodMappingExtractor.cs:127]: Added a test.

**Review by CommonGuy — APPROVED** [review]

**Pull request merged**